src/builder.ts

```diff
@@ -96,22 +105,25 @@ export default class Builder {
 						self.layers.push(sha)
 					}
 
-					this.emit('data', data.stream)
+					if (data.stream) {
+						outputStream.write(data.stream);
+					}
 				}
-			}))
+			})
+			.on('error', (error: Error) => outputStream.emit('error', error));
 
 			// Catch any errors the stream produces
-			outputStream.on('error', (err: Error) => {
-				self.callHook(hooks, 'buildFailure', handler, err, self.layers)
-			})
+			// dup.on('error') covers errors emitted by both inputStream and outputStream
 			dup.on('error', (err: Error) => {
-				self.callHook(hooks, 'buildFailure', handler, err, self.layers)
-			})
-
-			// Setup the buildSuccess hook. This handler is not called on
-			// error so we can use it to propagate the success information
-			outputStream.on('end', () => {
-				this.callHook(hooks, 'buildSuccess', handler, _.last(this.layers), this.layers)
+				errored = true;
+				self.callHook(hooks, 'buildFailure', handler, err, self.layers);
+			});
+
+			// Setup the buildSuccess hook
+			res.on('end', () => {
+				if (!errored) {
+					this.callHook(hooks, 'buildSuccess', handler, _.last(this.layers), this.layers);
```

Comment:
The `if` test is the fix for the bug of the build being occasionally reported successful when it actually failed.